pkg/util/tlsutil/tls.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2018 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tlsutil
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// CertType defines the type of the cert.
+type CertType int
+
+const (
+	// ClientCert defines a client cert.
+	ClientCert CertType = iota
+	// ServingCert defines a serving cert.
+	ServingCert
+	// ClientAndServingCert defines both client and serving cert.
+	ClientAndServingCert
+)
+
+// CertConfig configures how to generate the Cert.
+type CertConfig struct {
+	// CertName is the name of the cert.
+	CertName string
+	// Optional CertType. Serving, client or both; defaults to both.
+	CertType CertType
+	// Optional CommonName is the common name of the cert; defaults to "".
+	CommonName string
+	// Optional Organization is Organization of the cert; defaults to "".
+	Organization []string
+	// Optional CAKey, if user wants to provide custom CA key path.
+	CAKey string
+	// Optional CA Certificate, if user wants to provide custom CA cert path.
+	CACert string
```

Comment:
I have a concern that not all CAs are stored in this way. I would think a way to support multiple options by passing in the cert/key via path strings, and the user of the Operator would set up the mounts on the operator deployment?

Reply:
@shawn-hurley `I would think a way to support multiple options by passing in the cert/key via path strings`. 
Do you have a use case that prefers using cert file path on the operator pod over simply retrieving it via secret or ConfigMap. My initial philosophy on designing the operator util is the operator usually deals with Kubernete objects such as Pod, Deployment, Service, and so forth. Then it makes a lot of sense that the tls utility also returns Secret and Configmap so that the operator can just simply consume them without need for converting an external object into to a kubernete object.

Reply:
Offline Comments: 

I could also see that someone does not store the ca cert and specifically the key in the cluster and instead uses some other type of volume to store those values.

Reply:
Offline comments:

`Could also see that someone does not store the ca cert and specifically the key in the cluster and instead uses some other type of volume to store those values ie. NFS, aws, or azure for presistant storage.`

Reply:
@hasbro17 I agree with @shawn-hurley on making passing-in CA key and cert a bit more generic instead of forcing user to pass those in as secret/configmap. Suppose that the user mounts a aws volume that contain sensitive CA key, the user might not want to expose the key as secret into the kubernete cluster but simply want to use it to sign the private key and cert. Let me know what you think.

Reply:
Agreed. Using a secret and configmap would restrict users from accessing the CA through a mounted volume.

Let's just keep CAKey and CACert as mounted filepaths then.